In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('previous_application.csv')

In [ ]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [ ]:
df.shape

(1670214, 37)

In [ ]:
df.head()

,SK_ID_PREV,SK_ID_CURR,NAME_CONTRACT_TYPE,AMT_ANNUITY,AMT_APPLICATION,AMT_CREDIT,AMT_DOWN_PAYMENT,AMT_GOODS_PRICE,WEEKDAY_APPR_PROCESS_START,HOUR_APPR_PROCESS_START,FLAG_LAST_APPL_PER_CONTRACT,NFLAG_LAST_APPL_IN_DAY,RATE_DOWN_PAYMENT,RATE_INTEREST_PRIMARY,RATE_INTEREST_PRIVILEGED,NAME_CASH_LOAN_PURPOSE,NAME_CONTRACT_STATUS,DAYS_DECISION,NAME_PAYMENT_TYPE,CODE_REJECT_REASON,NAME_TYPE_SUITE,NAME_CLIENT_TYPE,NAME_GOODS_CATEGORY,NAME_PORTFOLIO,NAME_PRODUCT_TYPE,CHANNEL_TYPE,SELLERPLACE_AREA,NAME_SELLER_INDUSTRY,CNT_PAYMENT,NAME_YIELD_GROUP,PRODUCT_COMBINATION,DAYS_FIRST_DRAWING,DAYS_FIRST_DUE,DAYS_LAST_DUE_1ST_VERSION,DAYS_LAST_DUE,DAYS_TERMINATION,NFLAG_INSURED_ON_APPROVAL
0,2030495,271877,Consumer loans,1730.430,17145.0,17145.0,0.0,17145.0,SATURDAY,15,Y,1,0.0,0.182832,0.867336,XAP,Approved,-73,Cash through the bank,XAP,NaN,Repeater,Mobile,POS,XNA,Country-wide,35,Connectivity,12.0,middle,POS mobile with interest,365243.0,-42.0,300.0,-42.0,-37.0,0.0
1,2802425,108129,Cash loans,25188.615,607500.0,679671.0,NaN,607500.0,THURSDAY,11,Y,1,NaN,NaN,NaN,XNA,Approved,-164,XNA,XAP,Unaccompanied,Repeater,XNA,Cash,x-sell,Contact center,-1,XNA,36.0,low_action,Cash X-Sell: low,365243.0,-134.0,916.0,365243.0,365243.0,1.0
2,2523466,122040,Cash loans,15060.735,112500.0,136444.5,NaN,112500.0,TUESDAY,11,Y,1,NaN,NaN,NaN,XNA,Approved,-301,Cash through the bank,XAP,"Spouse, partner",Repeater,XNA,Cash,x-sell,Credit and cash offices,-1,XNA,12.0,high,Cash X-Sell: high,365243.0,-271.0,59.0,365243.0,365243.0,1.0
3,2819243,176158,Cash loans,47041.335,450000.0,470790.0,NaN,450000.0,MONDAY,7,Y,1,NaN,NaN,NaN,XNA,Approved,-512,Cash through the bank,XAP,NaN,Repeater,XNA,Cash,x-sell,Credit and cash offices,-1,XNA,12.0,middle,Cash X-Sell: middle,365243.0,-482.0,-152.0,-182.0,-177.0,1.0
4,1784265,202054,Cash loans,31924.395,337500.0,404055.0,NaN,337500.0,THURSDAY,9,Y,1,NaN,NaN,NaN,Repairs,Refused,-781,Cash through the bank,HC,NaN,Repeater,XNA,Cash,walk-in,Credit and cash offices,-1,XNA,24.0,high,Cash Street: high,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1670214 entries, 0 to 1670213
Data columns (total 37 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   SK_ID_PREV                   1670214 non-null  int64  
 1   SK_ID_CURR                   1670214 non-null  int64  
 2   NAME_CONTRACT_TYPE           1670214 non-null  object 
 3   AMT_ANNUITY                  1297979 non-null  float64
 4   AMT_APPLICATION              1670214 non-null  float64
 5   AMT_CREDIT                   1670213 non-null  float64
 6   AMT_DOWN_PAYMENT             774370 non-null   float64
 7   AMT_GOODS_PRICE              1284699 non-null  float64
 8   WEEKDAY_APPR_PROCESS_START   1670214 non-null  object 
 9   HOUR_APPR_PROCESS_START      1670214 non-null  int64  
 10  FLAG_LAST_APPL_PER_CONTRACT  1670214 non-null  object 
 11  NFLAG_LAST_APPL_IN_DAY       1670214 non-null  int64  
 12  RATE_DOWN_PAYMENT            774370 non-nu

In [ ]:
df.duplicated().sum()

np.int64(0)

Check SK_ID_CURR

In [ ]:
df.SK_ID_CURR.nunique()

338857

In [ ]:
df.SK_ID_CURR.is_unique

False

In [ ]:
df.SK_ID_CURR.value_counts().head()

,count
SK_ID_CURR,
187868,77
265681,73
173680,72
242412,68
206783,67


Check SK_ID_PERV

In [ ]:
df.SK_ID_PREV.nunique()

1670214

In [ ]:
df.SK_ID_PREV.is_unique

True

Check Categorical columns

In [ ]:
cat_col = df.select_dtypes(include='object').columns

In [ ]:
for col in cat_col:
  print(f"\n{'='* 60}")
  # print(col)
  print(df[col].value_counts(dropna=False).head())


NAME_CONTRACT_TYPE
Cash loans         747553
Consumer loans     729151
Revolving loans    193164
XNA                   346
Name: count, dtype: int64

WEEKDAY_APPR_PROCESS_START
TUESDAY      255118
WEDNESDAY    255010
MONDAY       253557
FRIDAY       252048
THURSDAY     249099
Name: count, dtype: int64

FLAG_LAST_APPL_PER_CONTRACT
Y    1661739
N       8475
Name: count, dtype: int64

NAME_CASH_LOAN_PURPOSE
XAP             922661
XNA             677918
Repairs          23765
Other            15608
Urgent needs      8412
Name: count, dtype: int64

NAME_CONTRACT_STATUS
Approved        1036781
Canceled         316319
Refused          290678
Unused offer      26436
Name: count, dtype: int64

NAME_PAYMENT_TYPE
Cash through the bank                        1033552
XNA                                           627384
Non-cash from your account                      8193
Cashless from the account of the employer       1085
Name: count, dtype: int64

CODE_REJECT_REASON
XAP       1353093
HC         

Missing percentages

In [ ]:
missing_pct = (
    df.isna()
    .mean()
    .mul(100)
    .sort_values(ascending=False)
)

missing_pct.head(15)

,0
RATE_INTEREST_PRIVILEGED,99.643698
RATE_INTEREST_PRIMARY,99.643698
AMT_DOWN_PAYMENT,53.636480
RATE_DOWN_PAYMENT,53.636480
NAME_TYPE_SUITE,49.119754
DAYS_TERMINATION,40.298129
DAYS_FIRST_DRAWING,40.298129
DAYS_FIRST_DUE,40.298129
DAYS_LAST_DUE_1ST_VERSION,40.298129
DAYS_LAST_DUE,40.298129


In [ ]:
df.AMT_DOWN_PAYMENT.value_counts().head()

,count
AMT_DOWN_PAYMENT,
0.0,369854
4500.0,21241
9000.0,14747
13500.0,9655
22500.0,8165


In [ ]:
df.RATE_DOWN_PAYMENT.value_counts().head()

,count
RATE_DOWN_PAYMENT,
0.000000,369854
0.108909,36341
0.217818,6482
0.326727,1081
0.544545,746


## **Understand all columns**


| Column | Decision | Reason |
| :--- | :--- | :--- |
| SK_ID_PREV | Use | Count/mapping ID, not a predictive value |
| SK_ID_CURR | Grouping key | Customer ID |
| NAME_CONTRACT_STATUS | Definitely use | Approved/refused history |
| NAME_CONTRACT_TYPE | Use | Previous loan types |
| AMT_APPLICATION | Use | Requested amount |
| AMT_CREDIT | Use | Granted credit |
| AMT_ANNUITY | Use | Payment burden |
| AMT_DOWN_PAYMENT | Use | Financing structure |
| CNT_PAYMENT | Use | Repayment term |
| DAYS_DECISION | Use | Recency of previous application |
| NAME_CLIENT_TYPE | Use | New/repeater customer behavior |
| NAME_PORTFOLIO | Use | Previous product portfolio |
| NAME_PRODUCT_TYPE | Use | Product/application type |
| CHANNEL_TYPE | Maybe use | Application channel |
| CODE_REJECT_REASON | Maybe use | Historical rejection reason |
| NAME_CASH_LOAN_PURPOSE | Low priority | Lots of XNA/XAP |
| NAME_PAYMENT_TYPE | Low priority | Less directly useful |
| NAME_TYPE_SUITE | Low priority | High missingness |
| RATE_INTEREST_PRIMARY | Skip initially | 99.64% missing |
| RATE_INTEREST_PRIVILEGED | Skip initially | 99.64% missing |
| DAYS_FIRST_DRAWING etc. | Skip initially | Missing ~40%, needs more interpretation |
| WEEKDAY_APPR_PROCESS_START | Skip initially | Little customer-level value |
| HOUR_APPR_PROCESS_START | Skip initially | Weak business relevance |


Aggregation

In [ ]:
# ============================================================
# 1. NUMERICAL / TIME-BASED AGGREGATION
# ============================================================

agg_rules = {

    # Number of previous applications
    'SK_ID_PREV': 'count',

    # Requested application amount
    'AMT_APPLICATION': ['sum', 'mean', 'max'],

    # Credit amount granted
    'AMT_CREDIT': ['sum', 'mean', 'max'],

    # Loan annuity / payment amount
    'AMT_ANNUITY': ['mean', 'max'],

    # Down payment
    'AMT_DOWN_PAYMENT': ['mean', 'max'],

    # Goods price
    'AMT_GOODS_PRICE': ['mean', 'max'],

    # Number of payments / loan term
    'CNT_PAYMENT': ['mean', 'max'],

    # Down payment rate
    'RATE_DOWN_PAYMENT': ['mean', 'max'],

    # How recently previous applications were decided
    'DAYS_DECISION': ['min', 'max', 'mean']
}


prev_all = (
    df
    .groupby('SK_ID_CURR')
    .agg(agg_rules)
)


prev_all.columns = [
    f'PREV_APP_{col}_{agg.upper()}'
    for col, agg in prev_all.columns
]

In [ ]:
# ============================================================
# 2. BUSINESS-FRIENDLY FEATURE NAMES
# ============================================================

prev_all = prev_all.rename(columns={

    'PREV_APP_SK_ID_PREV_COUNT':
        'PREV_APP_TOTAL_APPLICATIONS',

    'PREV_APP_AMT_APPLICATION_SUM':
        'PREV_APP_TOTAL_APPLICATION_AMOUNT',

    'PREV_APP_AMT_APPLICATION_MEAN':
        'PREV_APP_AVG_APPLICATION_AMOUNT',

    'PREV_APP_AMT_APPLICATION_MAX':
        'PREV_APP_MAX_APPLICATION_AMOUNT',

    'PREV_APP_AMT_CREDIT_SUM':
        'PREV_APP_TOTAL_CREDIT',

    'PREV_APP_AMT_CREDIT_MEAN':
        'PREV_APP_AVG_CREDIT',

    'PREV_APP_AMT_CREDIT_MAX':
        'PREV_APP_MAX_CREDIT',

    'PREV_APP_AMT_ANNUITY_MEAN':
        'PREV_APP_AVG_ANNUITY',

    'PREV_APP_AMT_ANNUITY_MAX':
        'PREV_APP_MAX_ANNUITY',

    'PREV_APP_AMT_DOWN_PAYMENT_MEAN':
        'PREV_APP_AVG_DOWN_PAYMENT',

    'PREV_APP_AMT_DOWN_PAYMENT_MAX':
        'PREV_APP_MAX_DOWN_PAYMENT',

    'PREV_APP_AMT_GOODS_PRICE_MEAN':
        'PREV_APP_AVG_GOODS_PRICE',

    'PREV_APP_AMT_GOODS_PRICE_MAX':
        'PREV_APP_MAX_GOODS_PRICE',

    'PREV_APP_CNT_PAYMENT_MEAN':
        'PREV_APP_AVG_PAYMENT_COUNT',

    'PREV_APP_CNT_PAYMENT_MAX':
        'PREV_APP_MAX_PAYMENT_COUNT',

    'PREV_APP_RATE_DOWN_PAYMENT_MEAN':
        'PREV_APP_AVG_DOWN_PAYMENT_RATE',

    'PREV_APP_RATE_DOWN_PAYMENT_MAX':
        'PREV_APP_MAX_DOWN_PAYMENT_RATE',

    'PREV_APP_DAYS_DECISION_MIN':
        'PREV_APP_OLDEST_DECISION',

    'PREV_APP_DAYS_DECISION_MAX':
        'PREV_APP_MOST_RECENT_DECISION',

    'PREV_APP_DAYS_DECISION_MEAN':
        'PREV_APP_AVG_DECISION_TIME'
})

In [ ]:
prev_all.head()

,PREV_APP_TOTAL_APPLICATIONS,PREV_APP_TOTAL_APPLICATION_AMOUNT,PREV_APP_AVG_APPLICATION_AMOUNT,PREV_APP_MAX_APPLICATION_AMOUNT,PREV_APP_TOTAL_CREDIT,PREV_APP_AVG_CREDIT,PREV_APP_MAX_CREDIT,PREV_APP_AVG_ANNUITY,PREV_APP_MAX_ANNUITY,PREV_APP_AVG_DOWN_PAYMENT,PREV_APP_MAX_DOWN_PAYMENT,PREV_APP_AVG_GOODS_PRICE,PREV_APP_MAX_GOODS_PRICE,PREV_APP_AVG_PAYMENT_COUNT,PREV_APP_MAX_PAYMENT_COUNT,PREV_APP_AVG_DOWN_PAYMENT_RATE,PREV_APP_MAX_DOWN_PAYMENT_RATE,PREV_APP_OLDEST_DECISION,PREV_APP_MOST_RECENT_DECISION,PREV_APP_AVG_DECISION_TIME
SK_ID_CURR,,,,,,,,,,,,,,,,,,,,
100001,1,24835.5,24835.50,24835.5,23787.0,23787.00,23787.0,3951.000,3951.000,2520.0,2520.0,24835.5,24835.5,8.0,8.0,0.104326,0.104326,-1740,-1740,-1740.0
100002,1,179055.0,179055.00,179055.0,179055.0,179055.00,179055.0,9251.775,9251.775,0.0,0.0,179055.0,179055.0,24.0,24.0,0.000000,0.000000,-606,-606,-606.0
100003,3,1306309.5,435436.50,900000.0,1452573.0,484191.00,1035882.0,56553.990,98356.995,3442.5,6885.0,435436.5,900000.0,10.0,12.0,0.050030,0.100061,-2341,-746,-1305.0
100004,1,24282.0,24282.00,24282.0,20106.0,20106.00,20106.0,5357.250,5357.250,4860.0,4860.0,24282.0,24282.0,4.0,4.0,0.212008,0.212008,-815,-815,-815.0
100005,2,44617.5,22308.75,44617.5,40153.5,20076.75,40153.5,4813.200,4813.200,4464.0,4464.0,44617.5,44617.5,12.0,12.0,0.108964,0.108964,-757,-315,-536.0


In [ ]:
# ============================================================
# 3. CONTRACT STATUS COUNTS
# ============================================================

status_counts = pd.crosstab(
    df['SK_ID_CURR'],
    df['NAME_CONTRACT_STATUS']
)

status_counts.columns = [
    f'PREV_APP_STATUS_{col.upper().replace(" ", "_")}_COUNT'
    for col in status_counts.columns
]

# Total applications
status_counts['PREV_APP_TOTAL_APPLICATIONS'] = (
    status_counts.sum(axis=1)
)

# Approval rate
status_counts['PREV_APP_APPROVAL_RATE'] = (
    status_counts['PREV_APP_STATUS_APPROVED_COUNT']
    / status_counts['PREV_APP_TOTAL_APPLICATIONS'].replace(0, np.nan)
)

# Refusal rate
status_counts['PREV_APP_REFUSAL_RATE'] = (
    status_counts['PREV_APP_STATUS_REFUSED_COUNT']
    / status_counts['PREV_APP_TOTAL_APPLICATIONS'].replace(0, np.nan)
)

status_counts.drop(columns='PREV_APP_TOTAL_APPLICATIONS',axis=1, inplace=True)


status_counts.head(2)

In [ ]:
# ============================================================
# 4. CONTRACT TYPE COUNTS
# ============================================================

contract_type_counts = pd.crosstab(
    df['SK_ID_CURR'],
    df['NAME_CONTRACT_TYPE']
)

contract_type_counts.columns = [
    f'PREV_APP_CONTRACT_{col.upper().replace(" ", "_")}_COUNT'
    for col in contract_type_counts.columns
]

contract_type_counts = contract_type_counts.drop(
    columns=['PREV_APP_CONTRACT_XNA_COUNT'],
    errors='ignore'
)

contract_type_counts.head(2)

In [ ]:
# ============================================================
# 5. CLIENT TYPE COUNTS
# ============================================================

client_type_counts = pd.crosstab(
    df['SK_ID_CURR'],
    df['NAME_CLIENT_TYPE']
)

client_type_counts.columns = [
    f'PREV_APP_CLIENT_{col.upper().replace(" ","_")}_COUNT'
    for col in client_type_counts.columns
]

client_type_counts = client_type_counts.drop(
    columns=['PREV_APP_CLIENT_XNA_COUNT'],
    errors='ignore'
)

client_type_counts.head(2)

,PREV_APP_CLIENT_NEW_COUNT,PREV_APP_CLIENT_REFRESHED_COUNT,PREV_APP_CLIENT_REPEATER_COUNT
SK_ID_CURR,,,
100001,0,1,0
100002,1,0,0


In [ ]:
# ============================================================
# 6. PORTFOLIO COUNTS
# ============================================================

portfolio_counts = pd.crosstab(
    df['SK_ID_CURR'],
    df['NAME_PORTFOLIO']
)

portfolio_counts.columns = [
    f'PREV_APP_PORTFOLIO_{col.upper()}_COUNT'
    for col in portfolio_counts.columns
]

portfolio_counts = portfolio_counts.drop(
    columns=['PREV_APP_PORTFOLIO_XNA_COUNT'],
    errors='ignore'
)

portfolio_counts.head(2)

,PREV_APP_PORTFOLIO_CARDS_COUNT,PREV_APP_PORTFOLIO_CARS_COUNT,PREV_APP_PORTFOLIO_CASH_COUNT,PREV_APP_PORTFOLIO_POS_COUNT
SK_ID_CURR,,,,
100001,0,0,0,1
100002,0,0,0,1


In [ ]:
# ============================================================
# 7. PRODUCT TYPE COUNTS
# ============================================================

product_type_counts = pd.crosstab(
    df['SK_ID_CURR'],
    df['NAME_PRODUCT_TYPE']
)

product_type_counts.columns = [
    f'PREV_APP_PRODUCT_{col.upper().replace("-", "_")}_COUNT'
    for col in product_type_counts.columns
]

product_type_counts = product_type_counts.drop(
    columns=['PREV_APP_PRODUCT_XNA_COUNT'],
    errors='ignore'
)

product_type_counts.head(2)

,PREV_APP_PRODUCT_WALK_IN_COUNT,PREV_APP_PRODUCT_X_SELL_COUNT
SK_ID_CURR,,
100001,0,0
100002,0,0


In [ ]:
# ============================================================
# 9. COMBINE ALL PREVIOUS-APPLICATION FEATURES
# ============================================================

final_previous_application = (
    prev_all
    .join(status_counts, how='outer')
    .join(contract_type_counts, how='outer')
    .join(client_type_counts, how='outer')
    .join(portfolio_counts, how='outer')
    .join(product_type_counts, how='outer')
)


# Missing values after joining (NAN)
final_previous_application = (
    final_previous_application
    .fillna(0)
)

# RESET INDEX
final_previous_application = (
    final_previous_application
    .reset_index()
)

final_previous_application.head()

,SK_ID_CURR,PREV_APP_TOTAL_APPLICATIONS,PREV_APP_TOTAL_APPLICATION_AMOUNT,PREV_APP_AVG_APPLICATION_AMOUNT,PREV_APP_MAX_APPLICATION_AMOUNT,PREV_APP_TOTAL_CREDIT,PREV_APP_AVG_CREDIT,PREV_APP_MAX_CREDIT,PREV_APP_AVG_ANNUITY,PREV_APP_MAX_ANNUITY,PREV_APP_AVG_DOWN_PAYMENT,PREV_APP_MAX_DOWN_PAYMENT,PREV_APP_AVG_GOODS_PRICE,PREV_APP_MAX_GOODS_PRICE,PREV_APP_AVG_PAYMENT_COUNT,PREV_APP_MAX_PAYMENT_COUNT,PREV_APP_AVG_DOWN_PAYMENT_RATE,PREV_APP_MAX_DOWN_PAYMENT_RATE,PREV_APP_OLDEST_DECISION,PREV_APP_MOST_RECENT_DECISION,PREV_APP_AVG_DECISION_TIME,PREV_APP_STATUS_APPROVED_COUNT,PREV_APP_STATUS_CANCELED_COUNT,PREV_APP_STATUS_REFUSED_COUNT,PREV_APP_STATUS_UNUSED_OFFER_COUNT,PREV_APP_APPROVAL_RATE,PREV_APP_REFUSAL_RATE,PREV_APP_CONTRACT_CASH_LOANS_COUNT,PREV_APP_CONTRACT_CONSUMER_LOANS_COUNT,PREV_APP_CONTRACT_REVOLVING_LOANS_COUNT,PREV_APP_CLIENT_NEW_COUNT,PREV_APP_CLIENT_REFRESHED_COUNT,PREV_APP_CLIENT_REPEATER_COUNT,PREV_APP_PORTFOLIO_CARDS_COUNT,PREV_APP_PORTFOLIO_CARS_COUNT,PREV_APP_PORTFOLIO_CASH_COUNT,PREV_APP_PORTFOLIO_POS_COUNT,PREV_APP_PRODUCT_WALK_IN_COUNT,PREV_APP_PRODUCT_X_SELL_COUNT
0,100001,1,24835.5,24835.50,24835.5,23787.0,23787.00,23787.0,3951.000,3951.000,2520.0,2520.0,24835.5,24835.5,8.0,8.0,0.104326,0.104326,-1740,-1740,-1740.0,1,0,0,0,1.0,0.0,0,1,0,0,1,0,0,0,0,1,0,0
1,100002,1,179055.0,179055.00,179055.0,179055.0,179055.00,179055.0,9251.775,9251.775,0.0,0.0,179055.0,179055.0,24.0,24.0,0.000000,0.000000,-606,-606,-606.0,1,0,0,0,1.0,0.0,0,1,0,1,0,0,0,0,0,1,0,0
2,100003,3,1306309.5,435436.50,900000.0,1452573.0,484191.00,1035882.0,56553.990,98356.995,3442.5,6885.0,435436.5,900000.0,10.0,12.0,0.050030,0.100061,-2341,-746,-1305.0,3,0,0,0,1.0,0.0,1,2,0,0,2,1,0,0,1,2,0,1
3,100004,1,24282.0,24282.00,24282.0,20106.0,20106.00,20106.0,5357.250,5357.250,4860.0,4860.0,24282.0,24282.0,4.0,4.0,0.212008,0.212008,-815,-815,-815.0,1,0,0,0,1.0,0.0,0,1,0,1,0,0,0,0,0,1,0,0
4,100005,2,44617.5,22308.75,44617.5,40153.5,20076.75,40153.5,4813.200,4813.200,4464.0,4464.0,44617.5,44617.5,12.0,12.0,0.108964,0.108964,-757,-315,-536.0,1,1,0,0,0.5,0.0,1,1,0,1,0,1,0,0,0,1,0,0


In [ ]:
# ============================================================
# 10. FINAL CHECK
# ============================================================

print("Shape:",
      final_previous_application.shape)

print(
    "Unique customers:",
    final_previous_application['SK_ID_CURR'].nunique())

print(
    "Duplicate customers:",
    final_previous_application['SK_ID_CURR'].duplicated().sum())

print("\nMissing values:")
display(
    final_previous_application
    .isna()
    .sum()
    .sort_values(ascending=False)
    .head(5)
)


Shape: (338857, 39)
Unique customers: 338857
Duplicate customers: 0

Missing values:


,0
SK_ID_CURR,0
PREV_APP_TOTAL_APPLICATIONS,0
PREV_APP_TOTAL_APPLICATION_AMOUNT,0
PREV_APP_AVG_APPLICATION_AMOUNT,0
PREV_APP_MAX_APPLICATION_AMOUNT,0


**SAVE**

In [ ]:
final_previous_application.to_csv(
    'previous_application_aggregated.csv',
    index=False
)